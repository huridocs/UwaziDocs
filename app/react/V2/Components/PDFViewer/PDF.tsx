import React, { Suspense, useEffect, useState } from 'react';
import loadable from '@loadable/component';
import { SelectionRegion, HandleTextSelection } from 'react-text-selection-handler';
import { TextSelection } from 'react-text-selection-handler/dist/TextSelection';
import { PDFDocumentProxy } from 'pdfjs-dist';
import { Translate } from 'app/I18N';
import { PDFJS, CMAP_URL } from './pdfjs';
import { TextHighlight } from './types';

const PDFPage = loadable(async () => import(/* webpackChunkName: "LazyLoadPDFPage" */ './PDFPage'));

interface PDFProps {
  fileUrl: string;
  highlights?: { [page: string]: TextHighlight[] };
  onSelect?: (selection: TextSelection) => any;
  onDeselect?: () => any;
}

const getPDFFile = async (fileUrl: string) =>
  PDFJS.getDocument({
    url: fileUrl,
    cMapUrl: CMAP_URL,
    cMapPacked: true,
  }).promise;

const PDF = ({ fileUrl, highlights, onSelect = () => {}, onDeselect }: PDFProps) => {
  const [pdf, setPDF] = useState<PDFDocumentProxy>();
  const [error, setError] = useState<string>();

  useEffect(() => {
    getPDFFile(fileUrl)
      .then(pdfFile => {
        setPDF(pdfFile);
      })
      .catch((e: Error) => {
        setError(e.message);
      });
  }, [fileUrl]);

  return error ? (
    <div>{error}</div>
  ) : (
    //@ts-ignore https://github.com/huridocs/uwazi/issues/6067
    <HandleTextSelection onSelect={onSelect} onDeselect={onDeselect}>
      <div id="pdf-container">
        {pdf &&
          Array.from({ length: pdf.numPages }, (_, index) => index + 1).map(number => {
            const page = number.toString();
            const pageHighlights = highlights ? highlights[page] : undefined;
            return (
              <Suspense key={`page-${page}`} fallback={<Translate>Loading</Translate>}>
                {/* @ts-ignore https://github.com/huridocs/uwazi/issues/6067 */}
                <SelectionRegion regionId={page}>
                  <PDFPage pdf={pdf} page={number} highlights={pageHighlights} />
                </SelectionRegion>
              </Suspense>
            );
          })}
      </div>
    </HandleTextSelection>
  );
};

export type { PDFProps };
export default PDF;
