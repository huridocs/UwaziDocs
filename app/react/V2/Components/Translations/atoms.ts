import { atom } from 'recoil';

const translationsAtom = atom({
  key: 'TranslationsAtom',
  default: [],
});

export { translationsAtom };
