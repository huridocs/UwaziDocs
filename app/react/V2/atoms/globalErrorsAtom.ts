import { atomWithReset } from 'jotai/utils';

const globalErrorsAtom = atomWithReset([]);

export { globalErrorsAtom };
