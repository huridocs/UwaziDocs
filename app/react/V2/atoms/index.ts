export { atomStore } from './store';
export { notificationAtom } from './notificationAtom';
export { settingsAtom } from './settingsAtom';
export { templatesAtom } from './templatesAtom';
export { translationsAtom } from './translationsAtom';
export { thesaurisAtom } from './thesaurisAtom';
export { globalMatomoAtom } from './globalMatomoAtom';
export { userAtom } from './userAtom';
export type { AtomStoreData } from './store';
export type { notificationAtomType } from './notificationAtom';
