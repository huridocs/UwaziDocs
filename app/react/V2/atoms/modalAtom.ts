import { atom } from 'recoil';
import { ModalProps } from '../Components/UI/Modal';

type modalAtomType = Omit<ModalProps, 'show'>;

const modalAtom = atom({
  key: 'Modal',
  default: {} as modalAtomType,
});

const showModalAtom = atom({
  key: 'ShowModal',
  default: false,
});

export type { modalAtomType };
export { modalAtom, showModalAtom };
