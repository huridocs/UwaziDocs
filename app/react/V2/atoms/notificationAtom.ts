import { atom } from 'recoil';
import { NotificationProps } from '../Components/UI/Notification';

type notificationAtomType = Omit<NotificationProps, 'dismissAction'>;

const notificationAtom = atom({
  key: 'Notification',
  default: {} as notificationAtomType,
});

export type { notificationAtomType };
export { notificationAtom };
