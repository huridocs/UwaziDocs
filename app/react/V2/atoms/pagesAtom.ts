import { atom } from 'recoil';
import { actions } from 'app/BasicReducer';
import { Page } from '../shared/types';

const pagesAtom = atom({
  key: 'pages',
  default: [] as Page[],
  //sync deprecated redux store
  effects: [
    ({ onSet }) => {
      onSet(pages => {
        actions.set('pages', pages);
      });
    },
  ],
});

export { pagesAtom };
