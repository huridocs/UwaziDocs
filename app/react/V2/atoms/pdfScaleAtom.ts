import { atom } from 'jotai';

const pdfScaleAtom = atom(1);

export { pdfScaleAtom };
