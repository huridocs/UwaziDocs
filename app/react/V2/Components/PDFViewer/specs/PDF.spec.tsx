/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, act, queryAllByAttribute, cleanup, RenderResult } from '@testing-library/react';
import { configMocks, mockIntersectionObserver } from 'jsdom-testing-mocks';
import { Provider } from 'react-redux';
import { pdfScaleAtom } from 'V2/atoms';
import { TestAtomStoreProvider, LEGACY_createStore as createStore } from 'V2/testing';
import PDF from '../PDF';

configMocks({ act });
const oberserverMock = mockIntersectionObserver();

const mockPageRender = jest.fn();
const mockPageDestroy = jest.fn();

const renderingStates = {
  INITIAL: 0,
  RUNNING: 1,
  PAUSED: 2,
  FINISHED: 3,
};

jest.mock('../pdfjs.ts', () => ({
  EventBus: jest.fn(),
  PDFJS: {
    getDocument: jest.fn(() => ({
      promise: Promise.resolve({
        numPages: 5,
        getPage: jest.fn(async () =>
          Promise.resolve({
            getViewport: () => ({ width: 100, height: 300 }),
          })
        ),
      }),
    })),
    PixelsPerInch: { PDF_TO_CSS_UNITS: 0.5 },
  },
  PDFJSViewer: {
    PDFPageView: jest.fn().mockImplementation(() => ({
      setPdfPage: jest.fn(),
      draw: jest.fn().mockImplementation(async () => {
        mockPageRender();
        return Promise.resolve();
      }),
      destroy: mockPageDestroy,
      renderingState: 0,
      cancelRendering: jest.fn(),
    })),
    RenderingStates: renderingStates,
  },
  CMAP_URL: 'legacy_character_maps',
}));

describe('PDF', () => {
  let renderResult: RenderResult;

  const renderComponet = () => {
    renderResult = render(
      <Provider store={createStore()}>
        <TestAtomStoreProvider initialValues={[[pdfScaleAtom, 1.5]]}>
          <PDF fileUrl="" />
        </TestAtomStoreProvider>
      </Provider>
    );
  };

  afterEach(() => {
    cleanup();
  });

  afterAll(() => {
    oberserverMock.cleanup();
  });

  it('should render the pdf file', async () => {
    await act(() => {
      renderComponet();
    });
    const { container } = renderResult;
    const page1 = queryAllByAttribute('class', container, 'pdf-page')[0];
    await act(() => {
      oberserverMock.enterNode(page1);
    });
    expect(mockPageRender).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
