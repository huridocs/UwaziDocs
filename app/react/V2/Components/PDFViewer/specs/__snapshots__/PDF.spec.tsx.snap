// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDF should render the pdf file 1`] = `
<div>
  <div
    role="none"
  >
    <div
      id="pdf-container"
      style="height: 100%; width: 100%; overflow: auto; padding-left: 10px; padding-right: 10px;"
    >
      <div
        id="page-1-container"
      >
        <div
          data-region-selector-id="1"
        >
          <div
            class="pdf-page"
          />
        </div>
      </div>
      <div
        id="page-2-container"
      >
        <div
          data-region-selector-id="2"
        >
          <div
            class="pdf-page"
          />
        </div>
      </div>
      <div
        id="page-3-container"
      >
        <div
          data-region-selector-id="3"
        >
          <div
            class="pdf-page"
          />
        </div>
      </div>
      <div
        id="page-4-container"
      >
        <div
          data-region-selector-id="4"
        >
          <div
            class="pdf-page"
          />
        </div>
      </div>
      <div
        id="page-5-container"
      >
        <div
          data-region-selector-id="5"
        >
          <div
            class="pdf-page"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
